### ISSUE TITLE
Opt out of disk caching for certain requests (or allow for encrypted disk caching)

### ISSUE THREAD

**Issue by SiavashB** [apollographql/apollo-kotlin#4721]
### Use case

We have a health care app that handles SOME hipaa secure material. If we store any patient information on disk, it has to be encrypted. We'd like to take advantage of disk caching for the rest of the feature, but there are a few requests within that feature that cant be stored without encryption. 
We already will have to have different Apollo clients for various features, we don't want to also create multiple clients per feature.

### Describe the solution you'd like

It would be nice if we could override the default caching set in the ApolloClient.Builder, such that we can opt out of disk caching on certain requests like this 

`apolloClient.query(Query()).fetchPolicy(CacheFirst).diskCaching(false)/.memoryCaching(false)`

So basically you would enable it on the Client, then have the option to desiable it on certain requests.

Alternatively our issue would be solved if we could encrypt the disk cache for these requests, but I know this kinda defeats the purpose since the decrypting process is slow

**Comment by martinbonnin** [comment]
You could design your own `CacheHeaders` and normalized cache that wraps the Sql one for this:

Define your cache:
```kotlin
  class MyCache(val delegate: NormalizedCache): NormalizedCache() {
    override fun merge(record: Record, cacheHeaders: CacheHeaders): Set<String> {
      if (cacheHeaders.hasHeader("DO_NO_STORE_ON_DISK")) {
        return emptySet()
      }
      return delegate.merge(record, cacheHeaders)
    }

    // other functions. There are a handful but none of them should contain much logic 
  }
```

And use it:

```kotlin
    val cacheFactory = MemoryCacheFactory().chain(object : NormalizedCacheFactory() {
      override fun create(): NormalizedCache {
        return MyCache(SqlNormalizedCache())
      }
    })
    apolloClient = ApolloClient.Builder().normalizedCache(cacheFactory).build()

    val cacheHeaders = CacheHeaders.builder().addHeader("DO_NO_STORE_ON_DISK", true).build()
    apolloClient.query(query).cacheHeaders(cacheHeaders).execute()
```

**Comment by martinbonnin** [comment]
@SiavashB anything else we can help with here?

**Comment by martinbonnin** [comment]
I'm going to assume you found a solution and close this issue. Let us know otherwise and we'll reopen it.
